diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel/conftest.sh nvidia-kmod-560.35.03-x86_64/kernel/conftest.sh
--- nvidia-kmod-560.35.03-x86_64.old/kernel/conftest.sh	2024-10-11 10:32:01.941882631 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel/conftest.sh	2024-10-11 10:32:08.270925396 +0200
@@ -6610,6 +6610,25 @@
             compile_check_conftest "$CODE" "NV_DRM_FBDEV_GENERIC_SETUP_PRESENT" "" "functions"
         ;;
 
+        drm_fbdev_ttm_setup)
+            #
+            # Determine whether drm_fbdev_ttm_setup is present.
+            #
+            # Added by commit 9060d7f49376 ("drm/fb-helper: Finish the
+            # generic fbdev emulation") in v4.19.
+            #
+            CODE="
+            #include <drm/drm_fb_helper.h>
+            #if defined(NV_DRM_DRM_FBDEV_TTM_H_PRESENT)
+            #include <drm/drm_fbdev_ttm.h>
+            #endif
+            void conftest_drm_fbdev_ttm_setup(void) {
+                drm_fbdev_ttm_setup();
+            }"
+
+            compile_check_conftest "$CODE" "NV_DRM_FBDEV_TTM_SETUP_PRESENT" "" "functions"
+        ;;
+
         drm_aperture_remove_conflicting_pci_framebuffers)
             #
             # Determine whether drm_aperture_remove_conflicting_pci_framebuffers is present.
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel/header-presence-tests.mk nvidia-kmod-560.35.03-x86_64/kernel/header-presence-tests.mk
--- nvidia-kmod-560.35.03-x86_64.old/kernel/header-presence-tests.mk	2024-10-11 10:32:01.941882631 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel/header-presence-tests.mk	2024-10-11 10:32:08.270925396 +0200
@@ -15,6 +15,7 @@
   drm/drm_atomic_uapi.h \
   drm/drm_drv.h \
   drm/drm_fbdev_generic.h \
+  drm/drm_fbdev_ttm.h \
   drm/drm_framebuffer.h \
   drm/drm_connector.h \
   drm/drm_probe_helper.h \
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-drv.c nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-drv.c
--- nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-drv.c	2024-10-11 10:32:01.939882618 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-drv.c	2024-10-11 10:32:08.270925396 +0200
@@ -73,6 +73,10 @@
 #include <drm/drm_fbdev_generic.h>
 #endif
 
+#if defined(NV_DRM_DRM_FBDEV_TTM_H_PRESENT)
+#include <drm/drm_fbdev_ttm.h>
+#endif
+
 #include <linux/pci.h>
 #include <linux/workqueue.h>
 
@@ -1851,7 +1855,11 @@
             drm_aperture_remove_conflicting_pci_framebuffers(pdev, nv_drm_driver.name);
 #endif
         }
+#if defined(NV_DRM_FBDEV_TTM_AVAILABLE)
+        drm_fbdev_ttm_setup(dev, 32);
+#else
         drm_fbdev_generic_setup(dev, 32);
+#endif
     }
 #endif /* defined(NV_DRM_FBDEV_GENERIC_AVAILABLE) */
 
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-os-interface.h nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-os-interface.h
--- nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-os-interface.h	2024-10-11 10:32:01.939882618 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-os-interface.h	2024-10-11 10:32:08.270925396 +0200
@@ -58,8 +58,11 @@
 #error "Need to define kernel timer callback primitives for this OS"
 #endif
 
-#if defined(NV_DRM_FBDEV_GENERIC_SETUP_PRESENT) && defined(NV_DRM_APERTURE_REMOVE_CONFLICTING_PCI_FRAMEBUFFERS_PRESENT)
+#if defined(NV_DRM_FBDEV_GENERIC_SETUP_PRESENT) || defined(NV_DRM_FBDEV_TTM_SETUP_PRESENT) && defined(NV_DRM_APERTURE_REMOVE_CONFLICTING_PCI_FRAMEBUFFERS_PRESENT)
 #define NV_DRM_FBDEV_GENERIC_AVAILABLE
+#if defined(NV_DRM_FBDEV_TTM_SETUP_PRESENT)
+#define NV_DRM_FBDEV_TTM_AVAILABLE
+#endif
 #endif
 
 struct page;
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-sources.mk nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-sources.mk
--- nvidia-kmod-560.35.03-x86_64.old/kernel/nvidia-drm/nvidia-drm-sources.mk	2024-10-11 10:32:01.940882624 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel/nvidia-drm/nvidia-drm-sources.mk	2024-10-11 10:32:08.270925396 +0200
@@ -67,6 +67,7 @@
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += sync_file_get_fence
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_aperture_remove_conflicting_pci_framebuffers
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_fbdev_generic_setup
+NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_fbdev_ttm_setup
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_connector_attach_hdr_output_metadata_property
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_helper_crtc_enable_color_mgmt
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_crtc_enable_color_mgmt
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel-open/conftest.sh nvidia-kmod-560.35.03-x86_64/kernel-open/conftest.sh
--- nvidia-kmod-560.35.03-x86_64.old/kernel-open/conftest.sh	2024-10-11 10:32:01.954882719 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel-open/conftest.sh	2024-10-11 10:32:27.985058605 +0200
@@ -6610,6 +6610,25 @@
             compile_check_conftest "$CODE" "NV_DRM_FBDEV_GENERIC_SETUP_PRESENT" "" "functions"
         ;;
 
+        drm_fbdev_ttm_setup)
+            #
+            # Determine whether drm_fbdev_ttm_setup is present.
+            #
+            # Added by commit 9060d7f49376 ("drm/fb-helper: Finish the
+            # generic fbdev emulation") in v4.19.
+            #
+            CODE="
+            #include <drm/drm_fb_helper.h>
+            #if defined(NV_DRM_DRM_FBDEV_TTM_H_PRESENT)
+            #include <drm/drm_fbdev_ttm.h>
+            #endif
+            void conftest_drm_fbdev_ttm_setup(void) {
+                drm_fbdev_ttm_setup();
+            }"
+
+            compile_check_conftest "$CODE" "NV_DRM_FBDEV_TTM_SETUP_PRESENT" "" "functions"
+        ;;
+
         drm_aperture_remove_conflicting_pci_framebuffers)
             #
             # Determine whether drm_aperture_remove_conflicting_pci_framebuffers is present.
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel-open/header-presence-tests.mk nvidia-kmod-560.35.03-x86_64/kernel-open/header-presence-tests.mk
--- nvidia-kmod-560.35.03-x86_64.old/kernel-open/header-presence-tests.mk	2024-10-11 10:32:01.954882719 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel-open/header-presence-tests.mk	2024-10-11 10:32:27.985058605 +0200
@@ -15,6 +15,7 @@
   drm/drm_atomic_uapi.h \
   drm/drm_drv.h \
   drm/drm_fbdev_generic.h \
+  drm/drm_fbdev_ttm.h \
   drm/drm_framebuffer.h \
   drm/drm_connector.h \
   drm/drm_probe_helper.h \
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-drv.c nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-drv.c
--- nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-drv.c	2024-10-11 10:32:01.951882699 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-drv.c	2024-10-11 10:32:27.985058605 +0200
@@ -73,6 +73,10 @@
 #include <drm/drm_fbdev_generic.h>
 #endif
 
+#if defined(NV_DRM_DRM_FBDEV_TTM_H_PRESENT)
+#include <drm/drm_fbdev_ttm.h>
+#endif
+
 #include <linux/pci.h>
 #include <linux/workqueue.h>
 
@@ -1851,7 +1855,11 @@
             drm_aperture_remove_conflicting_pci_framebuffers(pdev, nv_drm_driver.name);
 #endif
         }
+#if defined(NV_DRM_FBDEV_TTM_AVAILABLE)
+        drm_fbdev_ttm_setup(dev, 32);
+#else
         drm_fbdev_generic_setup(dev, 32);
+#endif
     }
 #endif /* defined(NV_DRM_FBDEV_GENERIC_AVAILABLE) */
 
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-os-interface.h nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-os-interface.h
--- nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-os-interface.h	2024-10-11 10:32:01.951882699 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-os-interface.h	2024-10-11 10:32:27.985058605 +0200
@@ -58,8 +58,11 @@
 #error "Need to define kernel timer callback primitives for this OS"
 #endif
 
-#if defined(NV_DRM_FBDEV_GENERIC_SETUP_PRESENT) && defined(NV_DRM_APERTURE_REMOVE_CONFLICTING_PCI_FRAMEBUFFERS_PRESENT)
+#if defined(NV_DRM_FBDEV_GENERIC_SETUP_PRESENT) || defined(NV_DRM_FBDEV_TTM_SETUP_PRESENT) && defined(NV_DRM_APERTURE_REMOVE_CONFLICTING_PCI_FRAMEBUFFERS_PRESENT)
 #define NV_DRM_FBDEV_GENERIC_AVAILABLE
+#if defined(NV_DRM_FBDEV_TTM_SETUP_PRESENT)
+#define NV_DRM_FBDEV_TTM_AVAILABLE
+#endif
 #endif
 
 struct page;
diff -Naur nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-sources.mk nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-sources.mk
--- nvidia-kmod-560.35.03-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-sources.mk	2024-10-11 10:32:01.952882705 +0200
+++ nvidia-kmod-560.35.03-x86_64/kernel-open/nvidia-drm/nvidia-drm-sources.mk	2024-10-11 10:32:27.985058605 +0200
@@ -67,6 +67,7 @@
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += sync_file_get_fence
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_aperture_remove_conflicting_pci_framebuffers
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_fbdev_generic_setup
+NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_fbdev_ttm_setup
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_connector_attach_hdr_output_metadata_property
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_helper_crtc_enable_color_mgmt
 NV_CONFTEST_FUNCTION_COMPILE_TESTS += drm_crtc_enable_color_mgmt
