diff -Naur nvidia-kmod-565.57.01-x86_64.old/kernel/nvidia-drm/nvidia-drm-drv.c nvidia-kmod-565.57.01-x86_64/kernel/nvidia-drm/nvidia-drm-drv.c
--- nvidia-kmod-565.57.01-x86_64.old/kernel/nvidia-drm/nvidia-drm-drv.c	2024-11-25 11:13:45.255224297 +0100
+++ nvidia-kmod-565.57.01-x86_64/kernel/nvidia-drm/nvidia-drm-drv.c	2024-11-25 11:14:14.771519760 +0100
@@ -1711,6 +1711,10 @@
     .read           = drm_read,
 
     .llseek         = noop_llseek,
+  
+#if defined(FOP_UNSIGNED_OFFSET)
+    .fop_flags   = FOP_UNSIGNED_OFFSET,
+#endif
 };
 
 static const struct drm_ioctl_desc nv_drm_ioctls[] = {
diff -Naur nvidia-kmod-565.57.01-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-drv.c nvidia-kmod-565.57.01-x86_64/kernel-open/nvidia-drm/nvidia-drm-drv.c
--- nvidia-kmod-565.57.01-x86_64.old/kernel-open/nvidia-drm/nvidia-drm-drv.c	2024-11-25 11:13:45.267224417 +0100
+++ nvidia-kmod-565.57.01-x86_64/kernel-open/nvidia-drm/nvidia-drm-drv.c	2024-11-25 11:14:02.253394455 +0100
@@ -1711,6 +1711,10 @@
     .read           = drm_read,
 
     .llseek         = noop_llseek,
+  
+#if defined(FOP_UNSIGNED_OFFSET)
+    .fop_flags   = FOP_UNSIGNED_OFFSET,
+#endif
 };
 
 static const struct drm_ioctl_desc nv_drm_ioctls[] = {
